ibis/expr/analysis.py

```diff
@@ -774,28 +776,15 @@ def predicate(node):
 
 def find_subqueries(node: ops.Node) -> Counter:
     counts = Counter()
-
-    def finder(node: ops.Node):
-        if isinstance(node, ops.Join):
-            return [node.left, node.right], None
-        elif isinstance(node, ops.PhysicalTable):
-            return g.halt, None
-        elif isinstance(node, ops.SelfReference):
-            return g.proceed, None
-        elif isinstance(node, (ops.Selection, ops.Aggregation)):
-            counts[node] += 1
-            return [node.table], None
-        elif isinstance(node, ops.TableNode):
-            counts[node] += 1
-            return g.proceed, None
-        elif isinstance(node, ops.TableColumn):
-            return node.table not in counts, None
-        else:
-            return g.proceed, None
-
-    # keep duplicates so we can determine where an expression is used
-    # more than once
-    list(g.traverse(finder, node, dedup=False))
+    for graph in map(g.Graph.from_dfs, util.promote_list(node)):
+        dependents = graph.invert()
```

Comment:
Why do we need to call `invert()` here? 
`toposort()` calculates the dependents as the first step.

Reply:
Quoting what I say above: 

> The number of unique dependents of the current node is the number of unique usages of the current subquery, which is used subsequently in the compiler to determine whether an expression should be compiled to a CTE.

Reply:
> toposort() calculates the dependents as the first step.

This is unrelated to the additional use of dependents.

Reply:
I played around with it a bit and now it makes total sense!

One thing I dislike is the additional `an.find_immediate_parent_tables`. I tried to remove it in https://github.com/ibis-project/ibis/pull/5575/commits/ac6592b954419220e266a8f5a4b5181f47d75433  and the only change is an additional CTE for where correlated subquery test.

Reply:
Yeah, that was there intentionally, to preserve as much of the original behavior of only counting a TableColumn dependent's child table once, but counting them more than once should not affect correctness so that SGTM!